libmcount/dynamic.c

```diff
@@ -327,12 +354,22 @@ static void prepare_dynamic_update(struct uftrace_sym_info *sinfo, bool needs_mo
 	};
 	int hash_size = sinfo->exec_map->mod->symtab.nr_sym * 3 / 4;
 
-	if (needs_modules)
+	if (code_hmap) {
+		/* main executable already loaded */
+		if (!needs_modules || modules_loaded)
+			return;
 		hash_size *= 2;
+	}
+	else { /* nothing loaded */
+		if (needs_modules)
+			hash_size *= 2;
+		code_hmap = hashmap_create(hash_size, hashmap_ptr_hash, hashmap_ptr_equals);
+	}
 
-	code_hmap = hashmap_create(hash_size, hashmap_ptr_hash, hashmap_ptr_equals);
+	int rc = dl_iterate_phdr(find_dynamic_module, &fmd);
```

Comment:
correct, but checking return code to be on the safe side. I am not sure about how dl_iterate_phdr is implemented. We will be able to catch a case where dl_iterate returns non-zero for some other reason not covered by the call back. Maybe that never happens, and may be it will?!

Reply:
The man page says it'd return the value from the last callback.